dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/RedshiftClusterUsageMetricsTask.java

```diff
@@ -0,0 +1,207 @@
+public class RedshiftClusterUsageMetricsTask extends AbstractAwsApiTask {
+
+  protected static enum MetricName {
+    CPUUtilization,
+    PercentageDiskSpaceUsed
+  }
+
+  protected static enum MetricType {
+    Average
+  }
+
+  @AutoValue
+  protected abstract static class MetricConfig {
+
+    public abstract MetricName name();
+
+    public abstract MetricType type();
+
+    public static MetricConfig create(MetricName name, MetricType type) {
+      return new AutoValue_RedshiftClusterUsageMetricsTask_MetricConfig(name, type);
+    }
+  }
+
+  @AutoValue
+  protected abstract static class MetricDataPoint {
+
+    public abstract Instant instant();
+
+    public abstract Double value();
+
+    public abstract MetricConfig metricConfig();
+
+    public static MetricDataPoint create(Instant instant, Double value, MetricConfig metricConfig) {
+      return new AutoValue_RedshiftClusterUsageMetricsTask_MetricDataPoint(
+          instant, value, metricConfig);
+    }
+  }
+
+  private static final String REDSHIFT_NAMESPACE = "AWS/Redshift";
+  private static final DateTimeFormatter DATE_FORMAT =
+      DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS").withZone(ZoneOffset.UTC);
+
+  private final ZonedDateTime currentTime;
+  private final ZonedInterval interval;
+  private final List<MetricConfig> metrics;
+
+  public RedshiftClusterUsageMetricsTask(
+      AWSCredentialsProvider credentialsProvider,
+      ZonedDateTime currentTime,
+      ZonedInterval interval,
+      String zipEntryName,
+      ImmutableList<MetricConfig> metrics) {
+    super(
+        credentialsProvider,
+        zipEntryName,
+        RedshiftRawLogsDumpFormat.ClusterUsageMetrics.Header.class);
+    this.interval = interval;
+    this.metrics = metrics;
+    this.currentTime = currentTime;
+  }
+
+  @Override
+  protected Void doRun(TaskRunContext context, @Nonnull ByteSink sink, Handle handle)
+      throws IOException {
+    writeRecordsCsv(
+        sink,
+        listClusters().stream()
+            .flatMap(
+                cluster ->
+                    getClusterMetrics(cluster.getClusterIdentifier()).entrySet().stream()
+                        .map(
+                            metricEntry ->
+                                serializeCsvRow(
+                                    cluster.getClusterIdentifier(),
+                                    metricEntry.getKey(),
+                                    metricEntry.getValue())))
+            .collect(toList()));
+    return null;
+  }
+
+  private Object[] serializeCsvRow(
+      String clusterId, Instant instant, List<MetricDataPoint> dataPoints) {
+    Map<MetricConfig, Double> values =
+        dataPoints.stream().collect(toMap(MetricDataPoint::metricConfig, MetricDataPoint::value));
+    return Stream.of(
+            new Object[] {clusterId, DATE_FORMAT.format(instant)},
+            metrics.stream().map(metric -> values.get(metric)).toArray())
```

Comment:
The code currently uses 2 steps to sort the data points: first, it collects the points to the intermediate map `values` and then it extracts the data points from the map.

I would suggest replacing this with the comparator in the constructor and then performing the mapping in the single step in this method:

```
private final Comparator<MetricDataPoint> dataPointsByMetricsComparator;

in ctor:
dataPointsByMetricsComparator = comparing(MetricDataPoint::metricConfig, Ordering.explicit(metrics));

in this method:
Object[] orderedDataPoints =
  dataPoints.stream().sorted(dataPointsByMetricsComparator).map(MetricDataPoint::value).toArray();
```
https://guava.dev/releases/snapshot/api/docs/com/google/common/collect/Ordering.html#explicit(java.util.List)

Reply:
My mistake. I didn't take into account that the input list might be incomplete.

In this case, the only optimization that I can see right now is reducing the two `toArray` invocations to one:
```
Stream.concat(Stream.of(clusterId, ...), metrics.stream().map(...)).toArray()
```